lib/puppet/type/rabbitmq_binding.rb

```diff
@@ -11,21 +11,68 @@
     end
   end
 
-  newparam(:name, :namevar => true) do
-    desc 'source and destination of bind'
-    newvalues(/^\S*@\S+@\S+$/)
+  # Match patterns without '@' as arbitrary names; match patterns with
+  # src@destination@vhost to their named params for backwards compatibility.
+  def self.title_patterns
+    [
+      [
+        /(^([^@]*)$)/m,
+        [
+          [ :name ]
+        ]
+      ],
+      [
+        /^(\S+)@(\S+)@(\S+)$/m,
+        [
+          [ :source ],
+          [ :destination],
+          [ :vhost ]
+        ]
+      ]
+    ]
+  end
+
+  newparam(:name) do
+    desc 'resource name, either source@destination@vhost or arbitrary name with params'
+
+    isnamevar
+  end
+
+  newparam(:source) do
```

Comment:
source/destination/vhost/routing_key should be properties because they can be read from & written to the system's configuration. This will also allow them to be printed when running `puppet resource rabbitmq_binding`